torch/csrc/autograd/engine.cpp

```diff
@@ -533,21 +542,62 @@ void GraphTask::exec_post_processing() {
   // added in between callback calls, so iterators may become invalidated.
   // NOLINTNEXTLINE(modernize-loop-convert)
-  for (size_t i = 0; i < final_callbacks_.size(); ++i) {
-    cb_lock.unlock();
-    final_callbacks_[i]();
-    cb_lock.lock();
+
+  // caller_current_streams_ with nullopt entries removed
+  std::vector<c10::Stream> caller_current_streams_filtered;
+
+  // See Note [Streaming backwards].
+  // Syncs caller_current_stream with leaf streams, so final_callbacks may use
+  // any grad on its device's current stream.
+  if (leaf_streams.size() > 0) {
+    for (const auto& leaf_stream : leaf_streams) {
+      // stash_current_streams() stashed streams for all device IDs that already had a
+      // CUDA context before the GraphTask executed. For inactive devices, it stashed
+      // a c10::nullopt. I don't expect GraphTask's backward pass ran leaf nodes on
+      // any new devices, so the stashed streams should be enough.
+      // If leaf_stream.device_index() happens to be for a new device,
+      // operator* on the c10::nullopt will throw an error.
+      const auto caller_current_stream = *caller_current_streams_[leaf_stream.device_index()];
+
+      if (caller_current_stream != leaf_stream) {
+        auto event = c10::Event{c10::DeviceType::CUDA};
+        event.record(leaf_stream);
+        caller_current_stream.wait(event);
+      }
+    }
+
+    caller_current_streams_filtered.reserve(caller_current_streams_.size());
+    for (const auto& opt_stream : caller_current_streams_) {
+      if (opt_stream.has_value()) {
+        caller_current_streams_filtered.push_back(*opt_stream);
+      }
+    }
   }
 
-  // Syncs leaf streams with default streams (if necessary)
-  // See note "Streaming backwards"
-  for (const auto& leaf_stream : leaf_streams) {
-    const auto guard = c10::impl::VirtualGuardImpl{c10::DeviceType::CUDA};
-    const auto default_stream = guard.getDefaultStream(leaf_stream.device());
-    if (leaf_stream != default_stream) {
+  {
+    // final_callbacks run on the per-device caller_current_streams (the ambient streams
+    // surrounding the user's call to backward()). This has two benefits:
+    //  1. caller_current_streams have been synced with leaf_streams, so callbacks may
+    //     safely access any grad.
+    //  2. The callback's results can safely be used on (user-facing) caller_current_streams
+    //     after backward().
+    c10::MultiStreamGuard g(caller_current_streams_filtered);
+    for (size_t i = 0; i < final_callbacks_.size(); ++i) {
+      cb_lock.unlock();
+      final_callbacks_[i]();
+      cb_lock.lock();
+    }
+  }
+
+  // For temporary BC, syncs default streams with caller_current_streams so callback results are also
+  // usable on user-facing default streams after backward()
+  for (const auto& caller_current_stream : caller_current_streams_filtered) {
+    const auto caller_default_stream = *caller_default_streams_[caller_current_stream.device_index()];
```

Comment:
You are dereferencing an optional without checking if it has value, this probably leads to test failure on ROCm